packages/taco/src/client/taco-client.ts

```diff
@@ -0,0 +1,291 @@
+ * const tacoClient = new TacoClient({
+ *   domain: DOMAIN_NAMES.TESTNET,
+ *   ritualId: 6,
+ *   ethersProvider,
+ *   ethersSigner
+ * });
+ *
+ * // Operations are safe and wait for readiness
+ * const messageKit = await tacoClient.encrypt('Hello, secret!', condition);
+ * const decrypted = await tacoClient.decrypt(messageKit, conditionContext);
+ * ```
+ */
+export class TacoClient {
+  private config: TacoClientConfig;
+  private static initializationPromise: Promise<void>;
+
+  /**
+   * Initialize TACo WASM globally (singleton pattern)
+   *
+   * This method ensures TACo WASM is initialized exactly once across all TacoClient instances.
+   * Initialization happens automatically when creating clients or calling operations, but you can
+   * call this explicitly for performance optimization or error handling.
+   *
+   * @returns {Promise<void>} Promise that resolves when TACo WASM is initialized
+   *
+   * @example
+   * ```typescript
+   * // Optional: Pre-initialize for better performance
+   * await TacoClient.initialize();
+   *
+   * // All TacoClient instances share the same initialization
+   * const client1 = new TacoClient(config1);
+   * const client2 = new TacoClient(config2);
+   *
+   * // Operations automatically wait for initialization
+   * const encrypted = await client1.encrypt(data, condition);
+   * ```
+   */
+  static async initialize(): Promise<void> {
+    if (!TacoClient.initializationPromise) {
+      TacoClient.initializationPromise = (async () => {
+        try {
+          await initialize();
+        } catch (error) {
+          console.error(`TACo initialization failed: ${error}`);
+          throw error; // Re-throw to maintain error propagation
+        }
+      })();
+    }
+    return TacoClient.initializationPromise;
+  }
+
+  /**
+   * Create a new TacoClient instance
+   *
+   * @param {TacoClientConfig} config - Configuration for the TacoClient
+   * @throws {Error} If configuration is invalid
+   */
+  constructor(config: TacoClientConfig) {
+    // Validate configuration using TacoConfig
+    const result = TacoConfigValidator.validateFast(config);
+    if (!result.isValid) {
+      throw new Error(`Invalid configuration: ${result.errors.join(', ')}`);
+    }
+
+    this.config = config;
+
+    TacoClient.initialize();
+  }
+
+  /**
+   * Fully validate the configuration including network provider checks
+   *
+   * This method performs comprehensive validation including:
+   * - Domain and ritual ID validation
+   * - Provider/signer configuration validation
+   * - Network compatibility check (calls provider to verify chain ID matches domain)
+   *
+   * @returns {Promise<ValidationResult>} Promise resolving to validation result with isValid boolean and errors array
+   * @throws {Error} If configuration validation fails
+   *
+   * @example
+   * ```typescript
+   * try {
+   *   const result = await tacoClient.validateConfig();
+   *   console.log('Configuration is valid:', result.isValid);
```

Comment:
This example has changed now that the function returns `Promise<Void>`, no?
```suggestion
   *   await tacoClient.validateConfig();
   *   console.log('Configuration is valid:', result.isValid);
```